crates/common/tedge_utils/src/fs_notify.rs

```diff
@@ -240,87 +242,104 @@ impl NotifyStream {
+                .watch_discriptor_id
+                .insert(wd.get_watch_descriptor_id(), full_path);
         }
         Ok(())
     }
 
     //// create an fs notification event stream
-    pub fn stream(mut self) -> impl Stream<Item = Result<(PathBuf, FileEvent), NotifyStreamError>> {
+    pub fn stream(self) -> impl Stream<Item = Result<(PathBuf, FileEvent), NotifyStreamError>> {
         try_stream! {
-            let mut notify_service = self.inotify.event_stream(self.buffer)?;
-            while let Some(event_or_error) = notify_service.next().await {
-                match event_or_error {
-                    Ok(event) => {
-                        let event_mask: FileEvent = event.mask.try_into()?;
-                        // because watching a file or watching a directory is implemented as
-                        // watching a directory, we can ignore the case where &event.name is None
-                        if let Some(event_name) = &event.name {
-                            let notify_file_name = event_name.to_str().ok_or_else(|| NotifyStreamError::FailedToCreateStream)?;
-                            // inotify triggered for a file named `notify_file_name`. Next we need
-                            // to see if we have a matching entry WITH a matching flag/mask in `self.watchers`
-                            for (dir_path, key) in &self.watchers.description {
-                                for (maybe_file_name, flags) in key {
-                                    for flag in flags {
-                                        // There are two cases:
-                                        // 1. we added a file watch
-                                        // 2. we added a directory watch
-                                        //
-                                        // for case 1. our input could have been something like:
-                                        // ...
+        let mut notify_service = self.inotify.into_event_stream(self.buffer)?;
+        while let Some(event_or_error) = notify_service.next().await {
+            match event_or_error {
+                Ok(event) => {
+                    let event_mask: FileEvent = event.mask.try_into()?;
+                    // because watching a file or watching a directory is implemented as
+                    // watching a directory, we can ignore the case where &event.name is None
+                    if let Some(event_name) = &event.name {
+                        let notify_file_name = event_name.to_str().ok_or_else(|| NotifyStreamError::FailedToCreateStream)?;
+                        // inotify triggered for a file named `notify_file_name`. Next we need
+                        // to see if we have a matching entry WITH a matching flag/mask in `self.watchers`
+                        for (dir_path, key) in &self.watchers.description {
+                            for (maybe_file_name, flags) in key {
+                                for flag in flags {
```

Comment:
This 3 levels loop is a sign that the data structure `watchers.description` is not adapted for the task. We are desperately looking for the file matching the event.  Now that a watch descriptor id is accessible from the inotify event, we can simply use a map that gives all what we need given such a watch descriptor id.